### ISSUE TITLE
git lfs pull a single file

### ISSUE THREAD

**Issue by mmahalwy** [github/git-lfs#1351]
Is there a way to only pull a single file?

**Comment by ttaylorr** [comment]
Hi @mmahalwy! You can specify the `--include` or `-I` flag (they are aliases of each other) to only include a specific filename in your pull.

For example, if you only wanted to pull the file called "a.dat", try:

```bash
$ git lfs pull --include "a.dat"
```

Or, if you only wanted to pull files matching the ".dat" extension, try:

```bash
$ git lfs pull --include "*.dat"
```

When given either the `--include` or `--exclude`, LFS will only pull files that are explicitly included and not excluded. For more information on these filters, you can check out our documentation [here](https://github.com/github/git-lfs/blob/7ee533f646ea7d76afcd073556c445bab420aa2b/docs/man/git-lfs-pull.1.ronn#L18-L24). 😄 

Let me know if you have any questions in the future!

**Comment by mmahalwy** [comment]
@ttaylorr thank you for that awesome answer! 

One last thing, does that include directories? For example, I have name `a.dat` that is inside directory `first/` and `second/`. I want to only pull `first/a.dat`.

Thoughts?

**Comment by ttaylorr** [comment]
Yep, just pass `-I first/a.dat`. `a.dat` alone will matcher neither `first/a.dat` or `second/a.dat`, since it pattern matches the whole path, not just the base 😄

**Comment by mmahalwy** [comment]
sweet! TY!

**Comment by ttaylorr** [comment]
You're welcome! Feel free to comment here again if you have any more questions 😄

**Comment by wutiejun** [comment]
I can use  ```git lfs pull -I "*a.dat"```, but ```git lfs pull -I "a.dat" ``` useless. Why?

**Comment by ttaylorr** [comment]
@wutiejun we have a test-case that is passing that illustrates this behavior working correctly [here](https://github.com/git-lfs/git-lfs/blob/v2.0.1/test/test-clone.sh#L334-L341).

If you're having trouble, feel free to open a new issue with further details.

**Comment by zezba9000** [comment]
How do you do this without a remote? Local repo only?

This doesn't work:

> git lfs checkout --include "image.bmp" > image.bmp.orig

**Comment by apm963** [comment]
@zezba9000 I could not find a native command (ala `show`) that returns the contents of an LFS object. Instead, I was able to get the file by building a script that performed the following steps:
*  Get output of `git show branch:path/to/file`. Ex:
   ```
   version https://git-lfs.github.com/spec/v1
   oid sha256:7de96d152019316b4b1d6c89979292fa5c4034c9f28bb3e3304717314ddea594
   size 279
   ```
*  Parse the output to get the hash after `sha256`. To do this I split the output into lines, then performed a regex match on the second line for something like `/^oid \w+:([a-fA-F0-9]+)$/`
*  Parse the hash for the following pieces
    *  Chars in position 0, 1 (eg. `7d`)
    *  Chars in position 2, 3 (eg. `e9`)
    *  The full hash
*  Using those facets, I attempt to get the file from the following path (relative to the repo root): `.git/lfs/objects/${firstHexPair}/${secondHexPair}/${fullHash}`

That worked well for me. You'll, of course, need to make sure that the LFS objects exist in your local repository.

**Comment by WilliamTambellini** [comment]
@apm963  thank you. 
Still, is it really the only way to extract the real file from a git lfs pointer file : should nt git lfs CLI provide a simple way to extract the file to a specified location ?

**Comment by bk2204** [comment]
You don't actually need a Git LFS-specific way to extract a file.  You can use Git to do it yourself by using `git cat-file --filters`, like so:

```sh
$ git cat-file --filters HEAD:test1.bin
```

**Comment by LeungMaggie** [comment]
git fetch
git checkout origin/branchxx path/to/file
